### DISCUSSION TITLE
Unterstützung Typ Sprinklersystem

### DISCUSSION THREAD

**Discussion by mbrockeu** [brutella/hkknx-public#281] [Ideas]
Ich habe gerade gesehen es gibt bei der Verwendung von Eve Aqua einen Typen "Sprinklersystem". 
Da ich aktuell meine Bewässerung mit Schaltaktor und Gardena gemacht habe wäre die Frage ob man den Typ direkt unterstützen kann. 

Aktuell nutze ich dafür einfach nur "Schalter" da es eh nur an und aus geht. Wäre aber für die Steuerung über Siri schöner wenn man sagen kann "Schalte die Bewässerung ein".

**Comment by brutella** [comment]
Ein Bewässerungssystem in HomeKit macht nur dann Sinn, wenn man die Beregnungszeit ändern kann. Ich selbst verwende ein "KNX-Bewässerungssystem" mit MDT Aktoren und Treppenlichtfunktion. Aber in meinem Fall würde es nicht funktionieren, da die Nachlaufzeit nicht über eine Gruppenadresse sondern nur über die ETS parametriert werden kann. 

Somit wäre der Vorteil, die Beregnungszeit einfach über die Home-App festzulegen, bei mir nicht gegeben.

Kannst du die Nachlaufzeit über eine GA ändern? Oder wie genau funktioniert dein Bewässerungssystem?

**Reply by mbrockeu** [reply]
Ich hatte gedacht HomeKit übernimmt die Nachlaufzeit. 

Aber ich habe auch gerade noch mal in die Beschreibung vom Schaltaktor geschaut. Ich nutze auch den MDT, aber aktuell nur an und aus. 

Es gibt aber neben treppenlicht auch treppenlicht mit Zeit. In der ETS kann ein Faktor definiert werden. Über die GA kommt dann der Wert den man als Treppenlicht nutzen möchte. 


Das sollte meiner Meinung nach dann doch gehen. Steht in Kapitel 4.5.7.

**Reply by brutella** [reply]
> Ich hatte gedacht HomeKit übernimmt die Nachlaufzeit.

Nein, das automatische Ausschalten müsste hkknx übernehmen. Das wäre an sich kein Problem. Aber ich würde die Nachlaufzeit dem MDT Aktor überlassen.

> Es gibt aber neben treppenlicht auch treppenlicht mit Zeit. In der ETS kann ein Faktor definiert werden. Über die GA kommt dann der Wert den man als Treppenlicht nutzen möchte.

Bei meinem Aktor AKK-0816.03 gibt es diese Funktion nicht. Wie sieht das bei dir aus?

Edit: Ah, die AKI/AKS-Serie kann das. Somit wäre das möglich.

**Reply by brutella** [reply]
Wie ich sehe kann man die voreingestellte Nachlaufzeit nicht auslesen. Da HomeKit diesen Wert aber wissen möchte, um die verbleibenden Zeit zu berechnen, müsste man dafür noch eine Lösung finden.

**Reply by mbrockeu** [reply]
Ich hab gerade auch festgestellt, ich hab da draussen ein AKK verbaut. Drinnen hab ich AKS verbaut. Da hätte man vermutlich über den Betriebsstundenzähler gehen können bzw. das Serviceintervall.

Andere alternative wäre halt über ne Logik z.b. vom Gira X1 und dort das Treppenhauslicht abzubilden. Aber so wie ich dich verstanden habe ist die Nachlaufzeit kein optionaler Parameter. Theoretisch hätte mir auch ein An/Aus gereicht.

Was ich mich aber frage, ich hatte in einem Video gesehen, dass die Sprinkler nacheinander geschaltet werden, dass würde doch nicht KNX machen sondern HomeKit, daher hatte ich auch angenommen der steuert die Zeit und sendet das Signal raus.

**Reply by brutella** [reply]
HomeKit zeigt nur die Daten an (Ventil ein/aus, verbleibende Beregnungsdauer) und schaltet von alleine nichts. Aus meiner Sicht gibt es folgende Möglichkeiten ein Bewässerungssystem einzubinden.

1. Mit einfachen Schaltern die Bewässerung manuell ein- und ausschalten. Mit optionaler Treppenlichtfunktion am Aktor würde sich die Bewässerung wieder automatisch abschalten.
2. Ein eigenes Bewässerungssystem in hkknx bauen. Die Ventile würden dann nach Ablauf einer gewissen Zeit von hkknx ausgeschaltet werden.

Mit Option 2 hätte man dann die Möglichkeit die Bewässerung zu automatisieren, entweder mit HomeKit-Automationen oder mit hkknx.
- HomeKit-Automationen: Damit könnte man die Bewässerung einschalten über die Apple-Home-App automatisieren – zB jeden Tag um 5.00 die Bewässerung einschalten.
- hkknx: Mit Skripte könnte einfache Zeitpläne oder auch komplexe Abläufe realisiert werden. Man könnte das Wetter berücksichtigen und auf KNX-Events reagieren. Als Alternative könnte hkknx ein eigenes Bewässerungssystem bereitstellen, was die selben Möglichkeiten wie Skripte hat, aber alles über die Weboberfläche konfigurierbar macht.

**Reply by mbrockeu** [reply]
Klingt auf jedenfalls nach einem großen Rattenschwanz der da dran hängt.

Mir würde Option 1 theoretisch reichen. Wenn man in der Konfiguration natürlich die Option hat GAs für Laufzeit und Restlaufzeit angeben zu können dann kann ja jeder seine Informationen da rein bringen.

Gibt es für die Funktionsweise in HomeKit irgendwo eine Beschreibung? So richtige Infos oder Videos wie genau das in HomeKit funktioniert findet sich ja nicht wirklich was ausser folgendes: https://youtu.be/aXRmxPpq07g?si=3OvQOy9Gs4xW0ssj&t=883

Deswegen bin ich irgendwie davon ausgegangen die logische Abfolge hintereinander kommt von HomeKit selbst.

**Reply by brutella** [reply]
> Wenn man in der Konfiguration natürlich die Option hat GAs für Laufzeit und Restlaufzeit angeben zu können dann kann ja jeder seine Informationen da rein bringen.

Kann man bei einem Aktor überhaupt die Restlaufzeit auslesen?

> Gibt es für die Funktionsweise in HomeKit irgendwo eine Beschreibung?

Im "HomeKit Accessory Protocol (HAP) Specification" Dokument unter 8.21 und 8.43 findet man etwas dazu.

<img width="889" alt="Screenshot 2023-09-05 at 10 48 44" src="https://github.com/brutella/hkknx-public/assets/125641/2f135aea-2be8-435c-8e13-d72e1d1b15d5">

<img width="905" alt="Screenshot 2023-09-05 at 10 48 05" src="https://github.com/brutella/hkknx-public/assets/125641/2d2f5366-ab06-4b81-a606-16355b17c0fb">

**Reply by mbrockeu** [reply]
Nein gibt es leider nicht. Aber muss man sich dann im Zweifel irgendwo berechnen. 
Das muss ja nicht zwangsweise in HKKNX sein. Würde das auch versuchen im X1 zu machen.

**Reply by brutella** [reply]
Wenn das nur über Umwege auslesbar ist, dann sollte das gleich hkknx machen. Ansonsten gibt es wieder Abhängigkeiten zu anderen Systemen.

Und das führt eigentlich dazu, dass hkknx ein eigenes Bewässerungssystem braucht. Da stellt sich für mich aber die Frage, wieviel Bedarf es an so einer Umsetzung gibt.

**Reply by mbrockeu** [reply]
Vermutlich eher gering. Daher die Idee mit der einfach Lösung. So hätte man zumindest den richtigen Typ in HomeKit. Momentan sind es nur Schalter. 

Das gute ist ja. Die Bewässerungssaison neigt sich dem Ende 😁 somit bleibt etwas Zeit für eine Implementierung oder um den Bedarf zu prüfen.

**Reply by brutella** [reply]
> Vermutlich eher gering. Daher die Idee mit der einfach Lösung. So hätte man zumindest den richtigen Typ in HomeKit. 

Wie du unter 8.43 nachlesen kannst, wird von HomeKit vorausgesetzt, dass wenn ein Ventil als Bewässerungsventil festgelegt wird (Valve Type is set to Irrigation), dann müssen die Eigenschaften "Set Duration" und "Remaining Duration" verfügbar sein. Da hätten wir schon das Problem.

Würde es genügen, einfach ein Ventil (oder Wasserhahn) zur Verfügung zu stellen?

**Reply by mbrockeu** [reply]
Wenn es sowas gibt würde das auch schon viel helfen :-)

Bzgl. Restlaufzeit mach ich mir mal Gedanken wie man das machen könnte, zumindest außerhalb HKKNX.

**Comment by Graefer** [comment]
> > Vermutlich eher gering. Daher die Idee mit der einfach Lösung. So hätte man zumindest den richtigen Typ in HomeKit.

👍
> Wie du unter 8.43 nachlesen kannst, wird von HomeKit vorausgesetzt, dass wenn ein Ventil als Bewässerungsventil festgelegt wird (Valve Type is set to Irrigation), dann müssen die Eigenschaften "Set Duration" und "Remaining Duration" verfügbar sein. Da hätten wir schon das Problem.

Interessant. 

Ich habe bei mir die Zeitfunktion des Irrigation systems noch nie genutzt, weil das zu komplex abläuft (variable Beregnungsdauer nach Zimmerman-Formel, dann sukzessive Reihenfolge, weil der Druck der Pumpe nur jeweils ein Ventil gleichzeitig bedienen kann). Dafür habe ich eine Logik. Wenn hkknx ein komplexes Beregnungsschema inkl. Zimmerman-Fsktor als (ggf. kostenpflichtige) Option leisten könnte, wäre das langfristig natürlich spannend. 

> Würde es genügen, einfach ein Ventil (oder Wasserhahn) zur Verfügung zu stellen?

Unbedingt. Ich habe ja schon oft um Unterstützung der Bewässerung gebeten. Denn ich möchte auch mal manuell ein Ventil öffnen. Und da wäre HomeKit eine schöne Hilfe. Der Vorteil gegenüber einfachem Schalter: man kann damit Bewässerungszonen erstellen (was bei vielen Ventilen Übersichtlichkeit schafft) und visuell darüber informieren, ob nur ein Ventil offen ist (linked valve), oder wirklich Wasser darüber läuft (ob die Pumpe als parent valve also in Betrieb ist oder eben nicht).

**Comment by mbrockeu** [comment]
> > Vermutlich eher gering. Daher die Idee mit der einfach Lösung. So hätte man zumindest den richtigen Typ in HomeKit.
> 
> Wie du unter 8.43 nachlesen kannst, wird von HomeKit vorausgesetzt, dass wenn ein Ventil als Bewässerungsventil festgelegt wird (Valve Type is set to Irrigation), dann müssen die Eigenschaften "Set Duration" und "Remaining Duration" verfügbar sein. Da hätten wir schon das Problem.
> 
> Würde es genügen, einfach ein Ventil (oder Wasserhahn) zur Verfügung zu stellen?

So die nächste Bewässerungssaison beginnt und wollte mal Fragen wie hier der Stand ist. Und sei es nur ein Ventil was als "Bewässerung" zu erkennen ist.

**Reply by brutella** [reply]
Kann aus meiner eigenen Erfahrung berichten, dass für mich eine Bewässerung über Schalter genügt. Jeder Bewässerungssektor hat einen eigenen Schalter, welcher ein 24V Ventil schaltet. Auch die Zisternenpumpe wird über einen Schalter angesteuert.

Der KNX Schaltaktor wurde so parametriert, dass ein Ventil automatisch nach 5 Minuten wieder ausschalten. Somit kann ich sicherstellen, dass ein Ventil nicht offen bleibt, weil entweder ich vergessen habe es wieder auszuschalten oder HomeKit gerade keine Verbindung zu meinem Gateway aufbauen kann. (Falls der KNX-Bus neu gestartet wird, weil der Strom ausgefallen ist, werden alle Ventile auf Aus gestellt.)

Die automatische Bewässerung mache ich dann über ein Skript, welches um 5:00 Uhr jeden Sektor nacheinander ein- und ausschaltet. Das Skript könnte man erweitern, dass vergangener oder vorausgesagter Regen berücksichtigt wird. Mach ich aber nicht, sondern schalte das Skript am Vorabend selbst aktiv oder eben nicht. 

Ich bewässere zurzeit nur meinen Rasen. Dieser wird nicht jeden Tag bewässert sondern 1 bis 2 mal pro Woche. An welchem Tag entscheide ich selbst.

Das Skript sieht in etwas so aus.

```go
var hkknx = import("hkknx")
var time = import("time")

morning = hkknx.AtTime(5, 0, 0)

for {
	select {
	case ←morning:
		hkknx.SendNotification("Bewässerung gestartet", "", "Matthias")
		// Zisternenpumpe einschalten
		hkknx.GroupWrite(true, "2/0/246")
		
		// Sektor C
		hkknx.GroupWrite(true, "2/0/242")
		<-time.After(5 * time.Minute)
		hkknx.GroupWrite(false, "2/0/242")
		
		 // Sektor A
		hkknx.GroupWrite(true, "2/0/240")
		<-time.After(5 * time.Minute)
		hkknx.GroupWrite(false, "2/0/240")
		
		// Sektor B
		hkknx.GroupWrite(true, "2/0/241")
		<-time.After(5 * time.Minute)
		hkknx.GroupWrite(false, "2/0/241")
		
		// Zisternenpumpe ausschalten
		hkknx.GroupWrite(false, "2/0/246")
		hkknx.SendNotification("Bewässerung beendet", "", "Matthias")
	}
}
```

Vorteile dieser Implementierung sind, dass 

- ich Benachrichtigungen senden kann, wenn die Bewässerung gestartet und beendet wurde. Somit habe ich am Morgen die Gewissheit, dass bewässert wurde.
- die KNX Schaltaktoren sicherstellen, dass keine Ventile geöffnet bleiben, falls etwas unvorhergesehenes passiert. zB Strom fällt aus, hkknx hat keine Verbindung mehr zum KNX-Bus, HomeKit zickt gerade rum.
- als Erweiterung die Wetterdaten abgefragt und für die Bewässerungslogik verwenden werde können
- die Bewässerungssektoren auch über KNX Taster ein- und ausschalten werden können.

Klar wird die Lösung nicht sauber als Bewässerung in HomeKit dargestellt, aber für mich ist es eine sehr gute und einfache Lösung.

Bin aber auch für alternative Ansätze offen. 😉

**Reply by mbrockeu** [reply]
Am Ende würde mir die Lösung so vermutlich auch reichen. Ich habe die jetzt auch über Treppenlicht angebunden. Ich habe ja letztens schon festgestellt, dass man bei einigen MDT Schaltaktoren die Zeit fürs Treppenlicht noch anpassen kann. Hab natürlich den falschen.

Ich fände es aber insgesamt schöner, wenn man in HomeKit den "Schalter" halt als Bewässerung geht. Wenn das funktionieren würde ohne den ganzen Kram ringsherum wäre das für mich ok. Klar ist am Ende nur ne optische Frage bzw. dann für die Steuerung über Siri.

**Reply by Graefer** [reply]
Für RemainingTime und SetDuratiion könnte man optional eine GA eintragen lassen bzw. ansonsten einfach einen Default-Wert setzen (hat ja keinen Einfluss auf den Status Active oder InUse und ist somit für die Funktion unerheblich).
